File: android/app/src/main/java/com/screentime/app/viewmodel/ActivityViewModel.kt
package com.screentime.app.viewmodel

import android.app.Application
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import androidx.core.content.getSystemService
import androidx.lifecycle.*
import com.screentime.app.data.model.*
import com.screentime.app.data.repository.RepositoryProvider
import com.screentime.app.notification.NotificationHelper
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlin.math.sqrt

class ActivityViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = RepositoryProvider.getRepository(application)
    private val sensorManager = application.getSystemService<SensorManager>()

    val allActivities: StateFlow<List<ActivityRecord>> = repository.allActivities
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val todayActivities: StateFlow<List<ActivityRecord>> = repository.getTodayActivities()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _currentActivity = MutableStateFlow<ActivityRecord?>(null)
    val currentActivity: StateFlow<ActivityRecord?> = _currentActivity

    private val _isVerifying = MutableStateFlow(false)
    val isVerifying: StateFlow<Boolean> = _isVerifying

    private val _verificationProgress = MutableStateFlow(0f)
    val verificationProgress: StateFlow<Float> = _verificationProgress

    private val _verificationMessage = MutableStateFlow("")
    val verificationMessage: StateFlow<String> = _verificationMessage

    private val _tapCount = MutableStateFlow(0)
    val tapCount: StateFlow<Int> = _tapCount

    private val _showRewardAnimation = MutableStateFlow(false)
    val showRewardAnimation: StateFlow<Boolean> = _showRewardAnimation

    private val _recentlyEarned = MutableStateFlow(0L)
    val recentlyEarned: StateFlow<Long> = _recentlyEarned

    private val requiredTaps = 20
    private var activityStartTime = 0L

    private val accelerometerListener = object : SensorEventListener {
        private var motionCount = 0
        override fun onSensorChanged(event: SensorEvent) {
            if (event.sensor.type == Sensor.TYPE_ACCELEROMETER) {
                val magnitude = sqrt(
                    event.values[0] * event.values[0] +
                    event.values[1] * event.values[1] +
                    event.values[2] * event.values[2]
                )
                if (magnitude > 12f) {
                    motionCount++
                    val progress = motionCount.toFloat() / (requiredTaps * 5)
                    _verificationProgress.value = minOf(1f, progress)
                    _verificationMessage.value = "Movement detected: ${(progress * 100).toInt()}%"
                    if (progress >= 1f) {
                        completeVerification(true)
                    }
                }
            }
        }
        override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    }

    fun startActivity(type: ActivityType) {
        val method = when (type) {
            ActivityType.WALKING, ActivityType.RUNNING, ActivityType.CYCLING,
            ActivityType.EXERCISE, ActivityType.OUTDOOR -> VerificationMethod.ACCELEROMETER
            ActivityType.READING -> VerificationMethod.SCROLL_DETECTION
            ActivityType.MEDITATION -> VerificationMethod.TAP_COUNT
            ActivityType.CUSTOM -> VerificationMethod.MANUAL
        }

        val activity = ActivityRecord(type = type, verificationMethod = method, status = ActivityStatus.IN_PROGRESS)
        _currentActivity.value = activity
        _isVerifying.value = true
        _tapCount.value = 0
        _verificationProgress.value = 0f
        activityStartTime = System.currentTimeMillis()

        viewModelScope.launch { repository.saveActivity(activity) }

        when (method) {
            VerificationMethod.TAP_COUNT -> _verificationMessage.value = "Tap the button $requiredTaps times to verify"
            VerificationMethod.SCROLL_DETECTION -> _verificationMessage.value = "Scroll through content to verify reading"
            VerificationMethod.ACCELEROMETER -> {
                startAccelerometerVerification()
                _verificationMessage.value = "Move your device to verify physical activity..."
            }
            VerificationMethod.MANUAL -> _verificationMessage.value = "Tap 'Complete' when done with your activity"
        }
    }

    fun recordTap() {
        if (!_isVerifying.value) return
        val newCount = _tapCount.value + 1
        _tapCount.value = newCount
        val progress = newCount.toFloat() / requiredTaps
        _verificationProgress.value = minOf(1f, progress)
        _verificationMessage.value = "$newCount/$requiredTaps interactions recorded"
        if (newCount >= requiredTaps) completeVerification(true)
    }

    fun recordScroll(distance: Float) {
        if (!_isVerifying.value) return
        val newProgress = minOf(1f, _verificationProgress.value + distance / 1000f)
        _verificationProgress.value = newProgress
        _verificationMessage.value = "Scroll progress: ${(newProgress * 100).toInt()}%"
        if (newProgress >= 1f) completeVerification(true)
    }

    fun completeManualActivity() = completeVerification(true)

    fun cancelActivity() {
        stopAccelerometer()
        _currentActivity.value?.let { activity ->
            viewModelScope.launch {
                repository.saveActivity(activity.copy(status = ActivityStatus.CANCELLED, endTime = System.currentTimeMillis()))
            }
        }
        resetVerification()
    }

    private fun completeVerification(success: Boolean) {
        stopAccelerometer()
        val activity = _currentActivity.value ?: return
        val duration = (System.currentTimeMillis() - activityStartTime) / 1000L

        val rewardSeconds = if (success) calculateReward(activity.type, duration) else 0L
        val updatedActivity = activity.copy(
            endTime = System.currentTimeMillis(),
            durationSeconds = duration,
            status = if (success) ActivityStatus.VERIFIED else ActivityStatus.FAILED,
            rewardEarnedSeconds = rewardSeconds
        )

        viewModelScope.launch {
            repository.saveActivity(updatedActivity)
            if (success) {
                _recentlyEarned.value = rewardSeconds
                _showRewardAnimation.value = true
                NotificationHelper.sendActivityCompletedNotification(
                    getApplication(),
                    activity.displayName,
                    rewardSeconds / 60
                )
                kotlinx.coroutines.delay(3000)
                _showRewardAnimation.value = false
            }
        }
        resetVerification()
    }

    private fun calculateReward(type: ActivityType, durationSeconds: Long): Long {
        val baseReward = type.rewardMinutes * 60
        val multiplier = minOf(durationSeconds.toDouble() / (15 * 60), 2.0)
        return (baseReward * multiplier).toLong()
    }

    private fun startAccelerometerVerification() {
        val accelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        accelerometer?.let {
            sensorManager.registerListener(accelerometerListener, it, SensorManager.SENSOR_DELAY_NORMAL)
        }
    }

    private fun stopAccelerometer() {
        sensorManager?.unregisterListener(accelerometerListener)
    }

    private fun resetVerification() {
        _isVerifying.value = false
        _currentActivity.value = null
        _tapCount.value = 0
        _verificationProgress.value = 0f
    }

    val todayEarnedSeconds: StateFlow<Long> = todayActivities.map { activities ->
        activities.filter { it.status == ActivityStatus.VERIFIED }.sumOf { it.rewardEarnedSeconds }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0L)

    override fun onCleared() {
        super.onCleared()
        stopAccelerometer()
    }
}

File: ios/ScreenTime/Sources/ViewModels/ActivityViewModel.swift
import Foundation
import Combine

class ActivityViewModel: ObservableObject {
    @Published var activities: [Activity] = []
    @Published var currentActivity: Activity?
    @Published var isVerifying: Bool = false
    @Published var verificationProgress: Double = 0
    @Published var verificationMessage: String = ""
    @Published var selectedActivityType: ActivityType = .walking
    @Published var recentlyEarned: TimeInterval = 0
    @Published var showRewardAnimation: Bool = false

    private var cancellables = Set<AnyCancellable>()
    private let dataStore = DataStore.shared
    private let verificationService = ActivityVerificationService.shared
    private let screenTimeService = ScreenTimeService.shared

    init() {
        setupBindings()
        activities = dataStore.activities
    }

    private func setupBindings() {
        dataStore.$activities
            .receive(on: DispatchQueue.main)
            .assign(to: &$activities)

        verificationService.$isVerifying
            .receive(on: DispatchQueue.main)
            .assign(to: &$isVerifying)

        verificationService.$verificationProgress
            .receive(on: DispatchQueue.main)
            .assign(to: &$verificationProgress)

        verificationService.$verificationMessage
            .receive(on: DispatchQueue.main)
            .assign(to: &$verificationMessage)
    }

    func startActivity(type: ActivityType) {
        let activity = Activity(
            type: type,
            startTime: Date(),
            verificationMethod: verificationMethodFor(type),
            status: .inProgress
        )
        currentActivity = activity
        dataStore.addActivity(activity)

        verificationService.startVerification(for: activity) { [weak self] success, duration in
            self?.completeActivity(success: success, duration: duration)
        }
    }

    func recordTap() {
        verificationService.recordTap()
    }

    func recordScroll(distance: Double) {
        verificationService.recordScroll(distance: distance)
    }

    func completeManualActivity() {
        verificationService.completeManualVerification()
    }

    func cancelActivity() {
        verificationService.cancelVerification()
        if let activity = currentActivity {
            var cancelled = activity
            cancelled.status = .cancelled
            cancelled.endTime = Date()
            dataStore.updateActivity(cancelled)
        }
        currentActivity = nil
    }

    private func completeActivity(success: Bool, duration: TimeInterval) {
        guard var activity = currentActivity else { return }
        activity.endTime = Date()
        activity.duration = duration
        activity.status = success ? .verified : .failed

        if success {
            let reward = verificationService.calculateReward(for: activity.type, duration: duration)
            activity.rewardEarned = reward
            recentlyEarned = reward
            screenTimeService.addEarnedTime(reward)
            showRewardAnimation = true
            NotificationService.shared.sendActivityCompletedNotification(
                activityName: activity.displayName,
                rewardMinutes: reward / 60
            )
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
                self.showRewardAnimation = false
            }
        }

        dataStore.updateActivity(activity)
        currentActivity = nil
    }

    private func verificationMethodFor(_ type: ActivityType) -> VerificationMethod {
        switch type {
        case .walking, .running, .cycling, .outdoor:
            return .accelerometer
        case .reading:
            return .scrollDetection
        case .meditation:
            return .tapCount
        case .exercise:
            return .accelerometer
        case .custom:
            return .manual
        }
    }

    var todayActivities: [Activity] {
        let today = Calendar.current.startOfDay(for: Date())
        return activities
            .filter { Calendar.current.isDate($0.startTime, inSameDayAs: today) }
            .sorted { $0.startTime > $1.startTime }
    }

    var todayEarned: TimeInterval {
        todayActivities
            .filter { $0.status == .verified }
            .reduce(0) { $0 + $1.rewardEarned }
    }
}
